File: Src/CSharp/Multiple/Xceed Windows Explorer/Utils/Icons/IconWrapper.cs
/*
 * Xceed Zip for .NET - Xceed Windows Explorer sample application
 * Copyright (c) 2006 - Xceed Software Inc.
 * 
 * [IconWrapper.cs]
 * 
 * This class wraps an icon to expose a custom hashcode for it.
 * 
 * This file is part of Xceed Zip for .NET. The source code in this file 
 * is only intended as a supplement to the documentation, and is provided 
 * "as is", without warranty of any kind, either expressed or implied.
 */

using System;
using System.Drawing;
using System.IO;

namespace Xceed.FileSystem.Samples.Utils.Icons
{
	public class IconWrapper
	{
    #region CONSTRUCTORS

		public IconWrapper( Icon icon )
		{
      if( icon == null )
        throw new ArgumentNullException( "icon" );

      try
      {
        // Create a byte array from the icon data.
        using( MemoryStream stream = new MemoryStream() )
        {
          icon.Save( stream );
          m_iconData = stream.ToArray();
        }
      }
      catch
      {
        m_iconData = new byte[ 0 ];
      }
    }

    #endregion CONSTRUCTORS

    #region PUBLIC PROPERTIES

    public byte[] IconData
    {
      get{ return m_iconData; }
    }

    #endregion PUBLIC PROPERTIES

    #region OVERRIDES

    public override bool Equals(object obj)
    {
      if( obj as IconWrapper == null )
        return false;

      byte[] targetArray = ( ( IconWrapper )obj ).IconData;

      if( m_iconData.Length != targetArray.Length )
        return false;

      for( int i = 0; i < m_iconData.Length; i++ )
      {
        if( m_iconData[ i ] != targetArray[ i ] )
          return false;
      }

      return true;
    }

    public override int GetHashCode()
    {
      int hash = 0;
      int index = 0;

      while( ( index + 4 ) < m_iconData.Length )
      {
        hash ^= BitConverter.ToInt32( m_iconData, index );
        index += 4;
      }
      
      return hash;
    }

    #endregion OVERRIDES

    #region PRIVATE FIELDS

    private byte[] m_iconData; //= null

    #endregion PRIVATE FIELDS
  }
}

File: Src/Visual Basic .NET/Multiple/Xceed Windows Explorer/Utils/Icons/IconWrapper.vb
'
 '* Xceed Zip for .NET - Xceed Windows Explorer sample application
 '* Copyright (c) 2006 - Xceed Software Inc.
 '*
'* [IconWrapper.vb]
 '*
 '* This class wraps an icon to expose a custom hashcode for it.
 '*
 '* This file is part of Xceed Zip for .NET. The source code in this file
 '* is only intended as a supplement to the documentation, and is provided
 '* "as is", without warranty of any kind, either expressed or implied.
 '


Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.IO

Namespace Xceed.FileSystem.Samples.Utils.Icons
  Public Class IconWrapper
    #Region "CONSTRUCTORS"

    Public Sub New(ByVal icon As Icon)
      If icon Is Nothing Then
        Throw New ArgumentNullException("icon")
      End If

      Try
        ' Create a byte array from the icon data.
        Dim stream As MemoryStream = New MemoryStream
        Try
          icon.Save(stream)
          m_iconData = stream.ToArray()
        Finally
          CType(stream, IDisposable).Dispose()
        End Try
      Catch
        m_iconData = New Byte() {}
      End Try
    End Sub

    #End Region ' CONSTRUCTORS

    #Region "PUBLIC PROPERTIES"

    Public ReadOnly Property IconData() As Byte()
      Get
        Return m_iconData
      End Get
    End Property

    #End Region ' PUBLIC PROPERTIES

    #Region "OVERRIDES"

    Public Overloads Overrides Function Equals(ByVal obj As Object) As Boolean
      If CType(IIf(TypeOf obj Is IconWrapper, obj, Nothing), IconWrapper) Is Nothing Then
        Return False
      End If

      Dim targetArray As Byte() = (CType(obj, IconWrapper)).IconData

      If m_iconData.Length <> targetArray.Length Then
        Return False
      End If

      Dim i As Integer = 0
      Do While i < m_iconData.Length
        If m_iconData(i) <> targetArray(i) Then
          Return False
        End If
        i += 1
      Loop

      Return True
    End Function

    Public Overrides Function GetHashCode() As Integer
      Dim hash As Integer = 0
      Dim index As Integer = 0

      Do While (index + 4) < m_iconData.Length
        hash = hash Xor BitConverter.ToInt32(m_iconData, index)
        index += 4
      Loop

      Return hash
    End Function

#End Region     ' OVERRIDES

    #Region "PRIVATE FIELDS"

    Private m_iconData As Byte() '= null

    #End Region ' PRIVATE FIELDS
  End Class
End Namespace
